client/src/Components/Layouts/Customers/editcustomer.jsx

```diff
@@ -0,0 +1,211 @@
+            }
+
+        openNotificationWithIcon = (type, message) => {
+            notification[type]({
+                message: message,
+                duration: 2
+            });
+        };
+        onCancel = e => {
+
+            this.props.changevisibility("customersPage", "edit", "editVisibility", [], '')(e);
+        }
+        onEdit = e => {
+            const form = this.props.form
+            form.validateFields((error, values) => {
+                if (error) {
+                    this.openNotificationWithIcon('error', error);
+                }
+                else {
+                    let editCustomer = {
+                        id: this.state.id,
+                        name: values.name,
+                        email: values.email,
+                        phone: parseInt(values.prefixPhone + values.phone),
+                        status: values.status==="true"?true:false,
+                        address: values.address,
+                        newpassword: values.newpassword,
+                        oldpassword: values.oldpassword
+                    }
+                    fetch(`api/v1/editCustomer/${this.props.id}`,
+                        {
+                            method: "PUT",
+                            headers: { 'content-type': 'application/json' },
+                            body: JSON.stringify(editCustomer)
+                        }).then(res => res.json()).then(result => {
+                            if (result.result) {
+                                this.openNotificationWithIcon('success', 'تم التعديل بنجاح');
+                                this.props.updateState(this.state.id,result.result)
+                                this.props.changevisibility("customersPage", "edit", "editVisibility", [], '')(e);
+                            } else {
+                                this.openNotificationWithIcon('error', result.error);
+                            }
+                        })
+                }
+            })
+        }
+
+        handelEmail = (rule, value, cb) => {
+            if (value.length) {
+                if (!validator.isEmail(value)) cb('يرجى ادخال قيمة صحيحة')
+                else cb()
+            } else cb('يرجى ادخال البريد الالكتروني')
+        }
+        handelName = (rule, value, cb) => {
+            if (!value) { cb('يرجى ادخال الاسم') }
+            else { cb() }
+        }
+        handlePhone = (rule, value, cb) => {
+            if (value) {
+
+                if (!value.match(/^[0-9]{9}$/)) {
+                    cb(' رقم الهاتف يجب ان يكون ارقام فقط وعددها 9 خانات')
+                }
+                else cb()
+            } else cb('يرجى ادخال رقم الهاتف')
+        }
+        handelState = (value) => {
+            if (value === true) return "true"
+            else return "false"
+        }
+        componentWillReceiveProps(props) {
+            const { visible, id, information } = props;
+            this.setState({
```

Comment:
@Jamalat-shamallakh You might be able to use these props directly in this component, instead of copying them into this.state.  This would be a better design, since copying props into state can create bugs (when they don't stay in sync) and causes extra renders (one render when the prop changes, then a second render when your component notices the change and updates the same value in its state).

Reply:
you are right :+1: